fix: Rework elasticsearch 7.x support to fix search issues

msavy opened the pull request.

msavy commented:
This is a rough cut right now. If it all works I'll tidy up the POMs, etc.

volkflo commented:
@msavy I start to test and review this tomorrow with fresh energy 🚀 Thanks! :)

msavy commented:
> @msavy I start to test and review this tomorrow with fresh energy 🚀 Thanks! :)

No rush :-)

msavy commented:
OK, I will tidy things tomorrow. Likely not worth spending time on this until we have time to talk.

I should do a quick investigation and we should discuss a couple of issues: 

Specifically, I am thinking of dcb3ff4 and similar cases where we need to add "keyword" to certain fields to make sorting work properly (NB: we can't actually change the test like this). 

I need to determine if there there is a way that we can we reliably tell when we need to use keyword for a random search query (and not break anything like dcb3ff4 does, because adding keyword to the test obviously ends up breaking non-ES stuff). 

One interesting solution I have thought of is that we can cache the ES schema we build and consult it to see whether that multifield has a 'keyword'; the current iteration code really has no clue about that, so we'd need some context so that we could look stuff up. 

BTW, I find it strange that you can't just ES to automatically use the keyword when using sorts/aggregations 🤔.

Anyway, I'll explain more tomorrow!

volkflo commented on manager/api/es/pom.xml:
```diff
@@ -43,6 +43,12 @@
+    <dependency>
+      <groupId>jakarta.annotation</groupId>
+      <artifactId>jakarta.annotation-api</artifactId>
+      <version>1.3.5</version>
```
Can we track this in the global `pom.xml`?

volkflo commented on manager/api/es/src/main/java/io/apiman/manager/api/es/EsStorage.java:
```diff
@@ -1688,9 +1748,9 @@ public int getMaxPolicyOrderIndex(String organizationId, String entityId, String
                 .fetchSource(fields, null)
                 .query(qb)
-                .sort(new FieldSortBuilder("name").order(SortOrder.ASC))
+                .sort(new FieldSortBuilder("name.keyword").order(SortOrder.ASC))
```
Are these okay or was this a test?

msavy replied:
Yeah, this is the final thing. Because a sort operation is being done here, we need to use the `.keyword` multifield (it's in the schema, so we can guarantee it will be there).

volkflo commented on manager/api/es/src/main/java/io/apiman/manager/api/es/EsStorage.java:
```diff
@@ -2682,15 +2742,15 @@ private QueryBuilder matchOrgQuery(String organizationId) {
-    protected List<String> getDefaultIndices() {
-        return Arrays.asList(EsConstants.MANAGER_INDEX_POSTFIXES);
-    }
+//
```
remove

volkflo commented on manager/api/war/tomcat8/src/main/java/io/apiman/manager/api/war/tomcat8/Tomcat8ApiManagerApplication.java:
```diff
@@ -1,29 +1,29 @@
-public class Tomcat8ApiManagerApplication extends ApiManagerApplication {
-
-}
+///**
```
Undo this, please :)

volkflo commented on manager/test/api/src/main/java/io/apiman/manager/test/junit/ManagerRestTester.java:
```diff
@@ -29,6 +29,7 @@
 import io.apiman.test.common.resttest.RestTest;
 import io.apiman.test.common.util.TestPlanRunner;
 import io.apiman.test.common.util.TestUtil;
+import java.io.IOException;
```
Is this used?

volkflo commented on manager/test/api/src/main/java/io/apiman/manager/test/server/ManagerApiTestServer.java:
```diff
@@ -180,8 +181,8 @@ protected void preStart() throws Exception {
     private static RestHighLevelClient createEsClient() {
-        Map<String, String> config = getTestClientConfig();
-        return new DefaultEsClientFactory().createClient(config, ES_DEFAULT_INDEX, getDefaultIndices());
+        Map<String, String> config = getTestClientConfig(); // EsStorage.indexDefinitions()
```
What does this comment mean? :)

msavy replied:
Just some artifact

volkflo commented on manager/test/api/src/test/java/io/apiman/manager/test/es/EsMetricsAccessorTest.java:
```diff
@@ -88,7 +89,18 @@ private static RestHighLevelClient createEsClient() {
-        return new DefaultEsClientFactory().createClient(config, APIMAN_METRICS_INDEX_NAME, getDefaultIndices());
+        config.put("client.type", "es");
+
+        // We want the metrics client to initialise its indexes and this is an easy way of doing that.
```
Please explain this :)

msavy replied:
The MetricsAccessorTest doesn't ever actually use the Metrics component to store its information. It uses a bulk upload method directly into the database. This means that the schema may not have been initialised.

To ensure that everything is done correctly, we start up the ESMetrics component, which will initialise its own schema.

If you look in the ESMetrics implementation you will see the corresponding (tiny) change that just ensure the client is initialised before it blocks.

volkflo commented on manager/test/api/src/test/resources/test-plan-data/searching/queries/misc/001_query-api-catalog.resttest:
```diff
@@ -1,41 +1,41 @@
-POST /search/apiCatalog/entries admin/admin
```
I think here is the line ending wrong?

msavy replied:
Yah, I think that came from Bastian's original commit. I'll back it out and push over the top of this commit unless someone tells me there's something we should have preserved in here.

volkflo commented on manager/test/api/src/test/resources/test-plan-data/searching/queries/misc/003_query-users.resttest:
```diff
@@ -3,7 +3,7 @@ Content-Type: application/json
 {
   "orderBy" : {
-    "name" : "fullName",
+    "name" : "fullName.keyword",
```
Was this a test?

msavy replied:
Yes, it was just an experiment

volkflo commented on manager/test/api/src/test/resources/test-plans/searching-testPlan.xml:
```diff
@@ -1,46 +1,69 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Line endings?

msavy replied:
I'll fix that. It came from the original commit I stacked on. BTW, here's where you can set it to auto-detect what line-endings to use, which removes all the pain in my experience!
![image](https://user-images.githubusercontent.com/423513/110147528-b305b800-7dd3-11eb-8296-67fd3f1a713e.png)

volkflo reviewed (changes requested):
Only 24 small comments :D
But it looks really nice now :)

msavy commented:
Thanks. I'm aware of those issues and this is just a draft I threw together that I will tidy before opening the pull request normally (see my previous comment explaining the details).

msavy commented:
Right, I've now squashed down all the work into a single commit and removed all of my test code and other experiments. 

It's ready for review now.

Thanks for your patience, guys.

I have added @BastianGem as `Co-Authored-By` for the whole commit to ensure he gets credit for his contributions on GH.

msavy commented on common/es/src/main/java/io/apiman/common/es/util/AbstractClientFactory.java:
```diff
@@ -74,14 +84,18 @@ protected void initializeIndices(RestHighLevelClient client, String indexPrefix,
                         boolean indexExists = client.indices().exists(indexExistsRequest, RequestOptions.DEFAULT);
                         if (!indexExists) {
-                            this.createIndex(client, indexPrefix, indexPostfix); //$NON-NLS-1$
+                            this.createIndex(client, index, indexPrefix, indexPostfix); //$NON-NLS-1$
```
I'm not sure if that gets used elsewhere 🤔. I'll check. I just wired it into existing code.

msavy commented on common/es/src/main/java/io/apiman/common/es/util/AbstractClientFactory.java:
```diff
@@ -133,29 +147,37 @@ private static boolean deleteIndex(RestHighLevelClient client, String indexName)
     /**
      * Creates an index.
      * @param client the elasticsearch client
+     * @param indexDef
```
Will fix

msavy commented on common/es/src/main/java/io/apiman/common/es/util/DefaultEsClientFactory.java:
```diff
@@ -63,34 +63,40 @@
+        String defaultIndexPrefix) {
+
         RestHighLevelClient client;
         String indexNamePrefix = config.get("client.indexPrefix"); //$NON-NLS-1$
```
Sounds a very good idea to me, but maybe we should do that as a separate mini-commit? This one is getting kinda big already and don't want to change too much extra stuff.

msavy replied:
https://github.com/apiman/apiman/issues/1025

msavy commented on common/es/src/main/java/io/apiman/common/es/util/EsConstants.java:
```diff
@@ -16,6 +16,8 @@
 
 import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
```
Looks like these should be removed

msavy commented on common/es/src/main/java/io/apiman/common/es/util/EsConstants.java:
```diff
@@ -78,72 +68,189 @@
 
+
+    // nested es fields
+    public static final String ES_NESTED_FIELD_API_POLICIES_POLICY_IMPL = ES_NESTED_FIELD_PREFIX_API_POLICIES + ES_FIELD_POLICY_IMPL;
```
We can leave this for another time, but I think this probs needs refactoring as well as I am not certain this is very clear. But it's fine for now!

msavy commented on common/es/src/test/java/io/apiman/common/es/util/builder/index/EsIndexPropertiesTest.java:
```diff
@@ -0,0 +1,235 @@
+        assertThatJson(result).isEqualTo(EXPECTED_INDEX_DEF_JSON);
+    }
+
+    private static final String EXPECTED_INDEX_DEF_JSON = "{\n"
```
Matter of taste I suppose 😅. I shall move it into a file.

msavy reviewed (commented):
OK all tidied up now. I also carved out @BastianGem's tests into a separate commit, and added coauthor credits for him. 

Any other refactors I think we should probably think about putting them into another commit.

Generally seems to work really nicely now, but let's give it a manual test before we give it the thumbs up.

Thanks, everyone.

msavy commented on gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java:
```diff
@@ -185,7 +185,7 @@ private HttpClientOptions configureHttpClientOptions(final URL artifactUrl) {
         } catch (UnknownHostException e) {
             //Non blocking error while trying to get no proxy addresses. Proxy will be always applied if configured.
-            e.printStackTrace();
+            //e.printStackTrace();
```
I have a separate ticket for this. The whole class is a bit of a mess and needs to be reworked. Check out https://github.com/apiman/apiman/issues/971 - (this was just to avoid obscuring what I was doing at the time and was not going to be in the non-draft commit)

volkflo commented on manager/test/api/src/test/resources/test-plan-data/searching/queries/apis/003_query-page-1.resttest:
```diff
@@ -0,0 +1,58 @@
+    }
+  ],
+  "totalSize":10
+}
```
Here we have probably a CLRF?

volkflo commented on manager/test/api/src/test/resources/test-plan-data/searching/queries/apis/004_query-page-2.resttest:
```diff
@@ -0,0 +1,58 @@
+    }
+  ],
+  "totalSize":10
+}
```
Here we have probably a CLRF?

volkflo commented on manager/test/api/src/test/resources/test-plans/searching-testPlan.xml:
```diff
@@ -38,9 +66,4 @@
     <test name="Invalid Filter Operator">test-plan-data/searching/validation/006_query-invalid-filter-operator.resttest</test>
   </testGroup>
 
-  <!-- How about some user testing? -->
```
Why do we delete this?

msavy replied:
@BastianGem this one is for you. I guess you felt there was enough user testing?

BastianGem replied:
I just moved this one to Misc lines 53-56

msavy commented on common/es/src/test/java/io/apiman/common/es/util/builder/index/EsIndexPropertiesTest.java:
```diff
@@ -0,0 +1,77 @@
+        assertThatJson(result).isEqualTo(resource("index-samples/OutputIndex.json"));
+    }
+
+}
```
I can't find any CRLF in this file. I've double and triple checked but don't seem to find it.

msavy commented on common/es/src/test/resources/index-samples/OutputIndex.json:
```diff
@@ -0,0 +1,124 @@
+      }
+    }
+  }
+}
```
Can't find any CRLF. All LF for me.

msavy commented:
Rebased.

volkflo commented:
@msavy I think the changes are okay now. @BastianGem can you check the index mapping once again because of the different fields for policies id.
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/13332383/110432052-f18ac380-80ae-11eb-8e7b-ca8116e733af.png)

msavy reviewed (commented):
LGTM now, thanks for your contribution to this one also @BastianGem

volkflo commented:
LGTM :) I just reviewed the changes that github marked as changed since the last review :)
Once this is merged we have to test carefully by hand if everything is working 🚀

msavy commented:
Probably just one file has a newline at the end and the other didn't. In
Unix-y tradition having a line break at the end of the file is typical and
most editors add it automatically.

On Tue, 9 Mar 2021, 13:48 Florian Volk, <notifications@github.com> wrote:

> LGTM :) I just reviewed the changes that github marked as changed since
> the last review :)
> Once this is merged we have to test carefully by hand if everything is
> working 🚀
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apiman/apiman/pull/1023#issuecomment-793922055>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADHMWIJECBWMTDK2QPF7V3TCYRMRANCNFSM4YTPCSIA>
> .
>

BastianGem reviewed (approved):
Looks good to me :)

msavy commented:
Thanks, guys. Took longer to get those last few little glitches ironed out, but seems good now.
